backend/session/logind.c

```diff
@@ -487,6 +487,115 @@ static int dbus_event(int fd, uint32_t mask, void *data) {
 	return 1;
 }
 
+static bool contains_str(char *needle, char **haystack) {
+  int i = 0;
+  while(haystack[i]) {
+    if(strcmp(haystack[i], needle) == 0) {
+      return true;
+    }
+
+    i++;
+  }
+
+  return false;
+}
+
+static bool get_display_session(char **session_id) {
+	assert(session_id != NULL);
+	int ret;
+
+	// If there's a session active for the current process then just use that
+	ret = sd_pid_get_session(getpid(), session_id);
+	if (ret == 0) {
+		return true;
+	}
+
+	// Find any active sessions for the user if the process isn't part of an
+	// active session itself
+	ret = sd_uid_get_display(getuid(), session_id);
+	if (ret < 0 && ret != -ENODATA) {
+		wlr_log(WLR_ERROR, "Failed to get display: %s", strerror(-ret));
+		return false;
+	} else if (ret < 0) {
+		char **user_sessions = NULL;
+
+		// Check if there are any greeter sessions available
+		int user_session_count = sd_uid_get_sessions(getuid(), 1, &user_sessions);
+		if (user_session_count < 0) {
+			wlr_log(WLR_ERROR, "Failed to get sessions");
+			return false;
+		}
+
+		if (user_session_count == 0) {
+			wlr_log(WLR_ERROR, "User has no sessions");
+			return false;
+		}
+
+		char *class = NULL;
+		for (int i = 0; i < user_session_count; ++i) {
+			ret = sd_session_get_class(user_sessions[i], &class);
+			if (ret < 0) {
+				continue;
+			}
+
+			if (strcmp(class, "greeter") == 0) {
+				*session_id = strdup(user_sessions[i]);
+				break;
+			}
+		}
+
+		for (int i = 0; i < user_session_count; ++i) {
+			free(user_sessions[i]);
+		}
+		free(user_sessions);
+		free(class);
+
+		if (!session_id) {
+			wlr_log(WLR_ERROR, "Couldn't find an active session or a greeter session");
+			return false;
+		}
+	}
```

Comment:
Since the `session_id` is passed into `get_display_session` I was thinking the free in the error handling of the calling function would take care of it: https://github.com/swaywm/wlroots/blob/bafaab3149212c72be8da9526d3935a017dbe0bd/backend/session/logind.c#L665

Is that not the case?

Reply:
Oh, I suppose that's right. I didn't look outside the function itself.
It's still weird for a function to leave a successfully allocated string inside the variable when it has failed, though. For consistency's sake, I'd still prefer it to be freed inside of the function and explicitly set `*session_id = NULL;` on failure.

Reply:
I've gone ahead and consolidated the de-allocations to the end of the function in the happy and error paths. It's a little awkward having the duplicated code freeing the same vars in the two different cases though. More then happy to hear any suggestions to clean it up

Reply:
I've refactored the frees to reduce duplication and am a bit more happy with it now